File: Templates/File Templates/AudioKit/AudioKit Instrument.xctemplate/___FILEBASENAME___.m
//
//  ___FILENAME___
//  ___PROJECTNAME___
//
//  Created by ___FULLUSERNAME___ on ___DATE___.
//___COPYRIGHT___
//

#import "___FILEBASENAME___.h"

@implementation ___FILEBASENAMEASIDENTIFIER___

- (instancetype)init
{
    self = [super init];
    if (self) {
        // Note Properties
        ___FILEBASENAMEASIDENTIFIER___Note *note = [[___FILEBASENAMEASIDENTIFIER___Note alloc] init];

        // Instrument Properties
        _amplitude = [self createPropertyWithValue:1.0 minimum:0.0 maximum:1.0];

        // Instrument Definition
        AKOscillator *oscillator = [AKOscillator oscillator];
        oscillator.frequency = note.frequency;
        oscillator.amplitude = note.amplitude;

        [self setAudioOutput:[oscillator scaledBy:_amplitude]];

        // Output to global effects processing (choose mono or stereo accordingly)
        _auxilliaryOutput = [AKAudio globalParameter];
        //_auxilliaryOutput = [AKStereoAudio globalParameter];
        [self assignOutput:_auxilliaryOutput to:oscillator];
    }
    return self;
}
@end

// -----------------------------------------------------------------------------
#  pragma mark - ___FILEBASENAMEASIDENTIFIER___ Note
// -----------------------------------------------------------------------------


@implementation ___FILEBASENAMEASIDENTIFIER___Note

- (instancetype)init
{
    self = [super init];
    if (self) {
        _frequency = [self createPropertyWithValue:440 minimum:100 maximum:20000];
        _amplitude = [self createPropertyWithValue:1 minimum:0 maximum:1];

        // Optionally set a default note duration
        self.duration.value = 1.0;
    }
    return self;
}

- (instancetype)initWithFrequency:(float)frequency amplitude:(float)amplitude;
{
    self = [self init];
    if (self) {
        _frequency.value = frequency;
        _amplitude.value = amplitude;
    }
    return self;
}

@end

File: Templates/File Templates/AudioKit/Swift AudioKit Instrument.xctemplate/___FILEBASENAME___.swift
//
//  ___FILENAME___
//  ___PROJECTNAME___
//
//  Created by ___FULLUSERNAME___ on ___DATE___.
//___COPYRIGHT___
//

class ___FILEBASENAME___: AKInstrument {

    // Instrument Properties
    var pan  = AKInstrumentProperty(value: 0.0, minimum: 0.0, maximum: 1.0)

    // Auxilliary Outputs (if any)
    var auxilliaryOutput = AKAudio()

    override init() {
        super.init()

        // Instrument Properties

        // Note Properties
        let note = ___FILEBASENAME___Note()

        // Instrument Definition
        let oscillator = AKFMOscillator();
        oscillator.baseFrequency = note.frequency;
        oscillator.amplitude = note.amplitude;

        let panner = AKPanner(audioSource: oscillator, pan: pan)

        auxilliaryOutput = AKAudio.globalParameter()
        assignOutput(auxilliaryOutput, to:panner)
    }
}

class ___FILEBASENAME___Note: AKNote {

    // Note Properties
    var frequency = AKNoteProperty(minimum: 440, maximum: 880)
    var amplitude = AKNoteProperty(minimum: 0,   maximum: 1)

    override init() {
        super.init()
        addProperty(frequency)
        addProperty(amplitude)
    }
}
